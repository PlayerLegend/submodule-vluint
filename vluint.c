#include <assert.h>
#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#define FLAT_INCLUDES
#include "../range/def.h"
#include "../window/def.h"
#include "../window/alloc.h"
#include "vluint.h"
#include "../log/log.h"

#define VLUINT_CHAR_MAX 128

unsigned long long vluint_read (bool * error, range_const_unsigned_char * input)
{
    unsigned long long result = 0;
    
    unsigned char c;

    unsigned long long power = 1;
    unsigned char cnum;

    while (true)
    {
	if (range_is_empty (*input))
	{
	    goto fail;
	}
	
	c = *(input->begin++);

	cnum = c & ~VLUINT_CHAR_MAX;

	result += power * cnum;

	if (c == cnum)
	{
	    return result;
	}

	power *= VLUINT_CHAR_MAX;
    }

fail:
    *error = true;
    return -1;
}

void vluint_write (window_unsigned_char * output, unsigned long long input)
{
    //buffer_rewrite (*output);

    while (input >= VLUINT_CHAR_MAX)
    {
	*window_push(*output) = (input % VLUINT_CHAR_MAX) | VLUINT_CHAR_MAX;
	input = input >> 7;

	assert (output->region.end[-1] & VLUINT_CHAR_MAX);
    }

    *window_push(*output) = input;
}
