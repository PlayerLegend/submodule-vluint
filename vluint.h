#ifndef FLAT_INCLUDES
#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#define FLAT_INCLUDES
#include "../range/def.h"
#include "../window/def.h"
#endif


void vluint_write (window_unsigned_char * output, unsigned long long input);
unsigned long long vluint_read (bool * error, range_const_unsigned_char * input);
