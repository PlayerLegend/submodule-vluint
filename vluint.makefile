test/vluint: \
	src/vluint/test/test.o \
	src/vluint/vluint.o \
	src/window/printf.o \
	src/window/printf_append.o \
	src/window/alloc.o \
	src/window/vprintf.o \
	src/window/vprintf_append.o


C_PROGRAMS += test/vluint
RUN_TESTS += test/vluint

vluint-tests: test/vluint

tests: vluint-tests

