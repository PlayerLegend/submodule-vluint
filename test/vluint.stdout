test int: 14578
length: 2
